File: ED Services/PluginVault/NetSuiteTokenGenerator.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Runtime.Serialization;
using System.ServiceModel;
using System.Text;
using System.Configuration;
using System.Security.Cryptography;
using Microsoft.Xrm.Sdk;

namespace EDServices
{
    
    public class NetSuiteTokenGenerator
    {
        public static SOAPTokenResponseType GetSOAPToken(SOAPTokenRequestType request, ITracingService tracingService, List<string> errorStack)
        {
            try
            {
                OAuthBase authBase = new OAuthBase();
                string timeStamp = authBase.GenerateTimeStamp();
                string nonce = authBase.GenerateNonce(tracingService, errorStack);


                string baseString = request.accountId + "&" + EDServicesHelper.EnvVariables["ts_NetSuiteConsumerKey"] + "&" + EDServicesHelper.EnvVariables["ts_NetSuiteTokenId"] + "&" + nonce + "&" + timeStamp;


                string consumerSecret = EDServicesHelper.EnvVariables["ts_NetSuiteConsumerSecret"];

                string tokenSecret = EDServicesHelper.EnvVariables["ts_NetSuiteTokenSecret"];

                string key = consumerSecret + "&" + tokenSecret;


                string signature = "";

                var encoding = new System.Text.ASCIIEncoding();

                byte[] keyByte = encoding.GetBytes(key);

                byte[] messageBytes = encoding.GetBytes(baseString);


                using (var myhmacsha256 = new HMACSHA256(keyByte))
                {
                    byte[] hashmessage = myhmacsha256.ComputeHash(messageBytes);
                    signature = Convert.ToBase64String(hashmessage);

                }

                SOAPTokenResponseType response = new SOAPTokenResponseType();
                response.nonce = nonce;
                response.timeStamp = timeStamp;
                response.signature = signature;


                return response;
            }
            catch (Exception e)
            {
                string error = "Error: " + e.Message;
                EDServicesHelper.writeToTrace("At GetSOAPToken(SOAPTokenRequestType request). " + error, tracingService
                    );
                errorStack.Add(error);
                return null;
            }

        }

        public static RESTTokenResponseType GetRESTSToken(RESTTokenRequestType request, ITracingService tracingService, List<string> errorStack)
        {
            try
            {
                OAuthBase authBase = new OAuthBase();
                string timeStamp = authBase.GenerateTimeStamp();
                string nonce = authBase.GenerateNonce(tracingService, errorStack);
                Uri url = new Uri(request.url);

                string norm = "";
                string norm1 = "";


                string consumerSecret = EDServicesHelper.EnvVariables["ts_NetSuiteConsumerSecret"];

                string tokenSecret = EDServicesHelper.EnvVariables["ts_NetSuiteTokenSecret"];

                string httpMethod = request.httpMethod;
                if (string.IsNullOrEmpty(httpMethod))
                {
                    httpMethod = "GET";
                }


                string signature = authBase.GenerateSignature256(url, request.consumerKey, consumerSecret, request.tokenId, tokenSecret, httpMethod, timeStamp, nonce, out norm, out norm1);

                if (signature.Contains("+"))
                {
                    signature = signature.Replace("+", "%2B");
                }

                string header = "Authorization: OAuth ";
                header += "oauth_signature=\"" + signature + "\",";
                header += "oauth_version=\"1.0\",";
                header += "oauth_nonce=\"" + nonce + "\",";
                header += "oauth_signature_method=\"HMAC-SHA256\",";
                header += "oauth_consumer_key=\"" + request.consumerKey + "\",";
                header += "oauth_token=\"" + request.tokenId + "\",";
                header += "oauth_timestamp=\"" + timeStamp + "\",";
                header += "realm=\"" + request.accountId + "\"";


                return new RESTTokenResponseType()
                {
                    authorization = header.Substring(15)
                };
            }
            catch (Exception e)
            {
                string error = "Error: " + e.Message;
                EDServicesHelper.writeToTrace("At GetRESTSToken(RESTTokenRequestType request). " + error, tracingService);
                errorStack.Add(error);
                return null;
            }
        }
    }

    [DataContract]
    public class SOAPTokenRequestType
    {
        [DataMember(Order = 0)]
        public string accountId { get; set; }

        [DataMember(Order = 1)]
        public string consumerKey { get; set; }

        [DataMember(Order = 2)]
        public string tokenId { get; set; }
    }

    [DataContract]
    public class RESTTokenRequestType
    {
        [DataMember(Order = 0)]
        public string accountId { get; set; }

        [DataMember(Order = 1)]
        public string consumerKey { get; set; }

        [DataMember(Order = 2)]
        public string tokenId { get; set; }

        [DataMember(Order = 3)]
        public string url { get; set; }

        [DataMember(Order = 4)]
        public string httpMethod { get; set; }
    }



    [DataContract]
    public class SOAPTokenResponseType
    {
        [DataMember(Order = 0, EmitDefaultValue = false)]
        public string nonce { get; set; }

        [DataMember(Order = 1, EmitDefaultValue = false)]
        public string timeStamp { get; set; }

        [DataMember(Order = 2, EmitDefaultValue = false)]
        public string signature { get; set; }

        [DataMember(Order = 3, EmitDefaultValue = false)]
        public string error { get; set; }
    }


    [DataContract]
    public class RESTTokenResponseType
    {
        [DataMember(Order = 0, EmitDefaultValue = false)]
        public string authorization { get; set; }

        [DataMember(Order = 3, EmitDefaultValue = false)]
        public string error { get; set; }
    }
}

File: Azure Function App Python/netsuite_token_generator.py
import hashlib
import hmac
import base64
from dataclasses import asdict, dataclass
import logging
import os
from typing import Optional
from urllib.parse import quote
import json

import techsoupservices_helper
from oauth_base import OAuthBase, SignatureTypes


@dataclass
class SOAPTokenRequest:
    account_id: str
    consumer_key: Optional[str] = None
    token_id: Optional[str] = None


@dataclass
class SOAPTokenResponse:
    nonce: str
    timestamp: str
    signature: str
    error: Optional[str] = None


@dataclass
class RESTTokenRequest:
    url: str
    account_id: Optional[str] = None
    consumer_key: Optional[str] = None
    token_id: Optional[str] = None    
    http_method: Optional[str] = "GET"


@dataclass
class RESTTokenResponse:
    authorization: str
    error: Optional[str] = None
    
    def to_dict(self):
        return asdict(self)
    
    def to_json(self):
        return json.dumps(asdict(self))


class NetSuiteTokenGenerator:

    @staticmethod
    def get_soap_token(logger: logging.Logger = None) -> Optional[SOAPTokenResponse]:
        try:
            auth_base = OAuthBase()

            timestamp = auth_base.generate_timestamp()
            nonce = techsoupservices_helper.get_netsuite_nonce(logger)

            # Get credentials from environment
            account_id = os.getenv("NetSuiteAccountId")
            consumer_key = os.getenv("NetSuiteConsumerKey")
            token_id = os.getenv("NetSuiteTokenId")
            consumer_secret = os.getenv("NetSuiteConsumerSecret")
            token_secret = os.getenv("NetSuiteTokenSecret")

            # Build base string for signature
            base_string = (
                f"{account_id}&"
                f"{consumer_key}&"
                f"{token_id}&"
                f"{nonce}&"
                f"{timestamp}"
            )

            # Build HMAC key
            key = f"{consumer_secret}&{token_secret}"

            # Generate signature using HMAC-SHA256
            key_bytes = key.encode("ascii")
            message_bytes = base_string.encode("ascii")

            hmac_obj = hmac.new(key_bytes, message_bytes, hashlib.sha256)
            signature = base64.b64encode(hmac_obj.digest()).decode("ascii")

            # Create response
            response = SOAPTokenResponse(
                nonce=nonce, timestamp=timestamp, signature=signature
            )

            return response

        except Exception as e:
            error = f"Error in get_soap_token: {str(e)}"
            if logger:
                logger.error(error)
            else:
                print(error)
            return SOAPTokenResponse(nonce="", timestamp="", signature="", error=error)

    @staticmethod
    def get_rest_token(
        request: RESTTokenRequest, logger: logging.Logger = None
    ) -> Optional[RESTTokenResponse]:
        try:
            if logger:
                logger.info(f"get_rest_token called")

            auth_base = OAuthBase()

            timestamp = auth_base.generate_timestamp()
            nonce = techsoupservices_helper.get_netsuite_nonce(logger)

            if not nonce:
                return RESTTokenResponse(
                    authorization="",
                    error="Failed to generate nonce"
                    )
            
            account_id = os.getenv("NetSuiteAccountId")
            consumer_key = os.getenv("NetSuiteConsumerKey")
            consumer_secret = os.getenv("NetSuiteConsumerSecret")
            token_id = os.getenv("NetSuiteTokenId")            
            token_secret = os.getenv("NetSuiteTokenSecret")

            # Default HTTP method to GET if not specified
            http_method = request.http_method if request.http_method else "GET"

            # Generate signature using HMAC-SHA256
            signature, normalized_url, normalized_params = (
                auth_base.generate_signature256(
                    url=request.url,
                    consumer_key=consumer_key,
                    consumer_secret=consumer_secret,
                    token=token_id,
                    token_secret=token_secret,
                    http_method=http_method,
                    timestamp=timestamp,
                    nonce=nonce,
                )
            )

            # URL encode the signature if it contains '+'
            if "+" in signature:
                signature = signature.replace("+", "%2B")

            # Build authorization header
            header = "OAuth "
            header += f'oauth_signature="{signature}",'
            header += 'oauth_version="1.0",'
            header += f'oauth_nonce="{nonce}",'
            header += 'oauth_signature_method="HMAC-SHA256",'
            header += f'oauth_consumer_key="{consumer_key}",'
            header += f'oauth_token="{token_id}",'
            header += f'oauth_timestamp="{timestamp}",'
            header += f'realm="{account_id}"'

            response = RESTTokenResponse(authorization=header)

            return response

        except Exception as e:
            error = f"Error in get_rest_token: {str(e)}"
            if logger:
                logger.error(error)
            else:
                print(error)
            return RESTTokenResponse(
                authorization="",
                error=error
                )


def generate_soap_token() -> Optional[SOAPTokenResponse]:
    return NetSuiteTokenGenerator.get_soap_token()


def generate_rest_token(
    url: str, http_method: str = "GET"
) -> Optional[RESTTokenResponse]:

    request = RESTTokenRequest(url=url, http_method=http_method)
    return NetSuiteTokenGenerator.get_rest_token(request)
